### PR TITLE
Test improvisation for Superfluid

### PR DESCRIPTION
**Pull request by mattverse**
## Description

This PR improvises superfluid testing environment in the following ways:
- Add tests for IntermediaryAccount, IntermediaryAccountConnection
-  Add test packages for balancer and lockup so that they have publicly exposed methods for suite (cref: https://github.com/osmosis-labs/osmosis/issues/1033)
- Improvise hooks test so that it covers superfluid flow test
- add minor comment on `gauge.proto`
______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: x/superfluid/keeper/hooks_test.go [hunk 1]
```diff
@@ -18,7 +16,7 @@ func (suite *KeeperTestSuite) TestSuperfluidAfterEpochEnd() {
 			"happy path with single validator and delegator",
 			[]stakingtypes.BondStatus{stakingtypes.Bonded},
 			[]superfluidDelegation{{0, 0, 0, 1000000}},
-			sdk.Coins{},
+			sdk.Coins{{Amount: sdk.NewInt(999990), Denom: "stake"}},
```
[the remark this answers is not in the document]
Comment (mattverse) [reply x/superfluid/keeper/hooks_test.go:19]: It's to test distribution rewards, do you think I should try using some other way to test distributions?

File: x/superfluid/keeper/stake_test.go [hunk 2]
```diff
@@ -86,7 +86,7 @@ func (suite *KeeperTestSuite) SetupSuperfluidDelegate(delAddr sdk.AccAddress, va
 	coins := sdk.Coins{sdk.NewInt64Coin(denom, amount)}
 	lastLockID := suite.App.LockupKeeper.GetLastLockID(suite.Ctx)
 
-	lockID := suite.LockTokens(delAddr, coins, unbondingDuration)
+	lockID := suite.lockupTestHelper.LockTokens(delAddr, coins, unbondingDuration)
```
[the remark this answers is not in the document]
Comment (mattverse) [reply x/superfluid/keeper/stake_test.go:89]: Do you mean I should change other tests in other packages to use this as well?

File: app/apptesting/test_suite.go [hunk 3]
```diff
@@ -99,3 +106,100 @@ func (keeperTestHelper *KeeperTestHelper) EndBlock() {
 	reqEndBlock := abci.RequestEndBlock{Height: keeperTestHelper.Ctx.BlockHeight()}
 	keeperTestHelper.App.EndBlocker(keeperTestHelper.Ctx, reqEndBlock)
 }
+
+func (keeperTestHelper *KeeperTestHelper) AllocateRewardsToValidator(valAddr sdk.ValAddress) {
```
Comment (ValarDragon) [inline app/apptesting/test_suite.go:110]: Can we add the reward amount as an argument here

File: app/apptesting/test_suite.go [hunk 4]
```diff
@@ -99,3 +106,100 @@ func (keeperTestHelper *KeeperTestHelper) EndBlock() {
...
+	// allocate reward tokens to distribution module
+	coins := sdk.Coins{sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(20000))}
+	err := simapp.FundModuleAccount(keeperTestHelper.App.BankKeeper, keeperTestHelper.Ctx, distrtypes.ModuleName, coins)
+	keeperTestHelper.Require().NoError(err)
+
+	// allocate rewards to validator
+	keeperTestHelper.Ctx = keeperTestHelper.Ctx.WithBlockHeight(keeperTestHelper.Ctx.BlockHeight() + 1)
+	decTokens := sdk.DecCoins{{Denom: sdk.DefaultBondDenom, Amount: sdk.NewDec(20000)}}
+	keeperTestHelper.App.DistrKeeper.AllocateTokensToValidator(keeperTestHelper.Ctx, validator, decTokens)
+	keeperTestHelper.App.DistrKeeper.IncrementValidatorPeriod(keeperTestHelper.Ctx, validator)
```
Comment (ValarDragon) [inline app/apptesting/test_suite.go:123]: why is this changing the validator period? I don't think it should be, the validators delegation amount isn't changing

File: x/superfluid/keeper/hooks_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/superfluid/keeper/hooks_test.go:19]: We should have some working out of what we expect this number to be, I think its ok to mark this as a TODO, and making good test cases for reward distribution a TODO, and fix them both in a follow up PR if thats easiest.

File: app/apptesting/test_suite.go (hunk 4 above)
Comment (mattverse) [reply app/apptesting/test_suite.go:123]: Oh is it ran upon delegation amount change? I thought it was running based on reward change on validators

Comment (ValarDragon) [reply app/apptesting/test_suite.go:123]: nope, delegation amount change only!

File: app/apptesting/test_suite.go [hunk 5]
```diff
@@ -99,3 +106,100 @@ func (keeperTestHelper *KeeperTestHelper) EndBlock() {
...
+}
+
+func (keeperTestHelper *KeeperTestHelper) SetupGammPoolsWithBondDenomMultiplier(multipliers []sdk.Dec) []gammtypes.PoolI {
+	keeperTestHelper.App.GAMMKeeper.SetParams(keeperTestHelper.Ctx, gammtypes.Params{
+		PoolCreationFee: sdk.Coins{},
+	})
```
Comment (ValarDragon) [inline app/apptesting/test_suite.go:129]: can we instead mint the pool creation fee tokens to the creating address? (We can do this via `simapp.FundAccount`)

File: app/apptesting/test_suite.go [hunk 6]
```diff
@@ -99,3 +106,100 @@ func (keeperTestHelper *KeeperTestHelper) EndBlock() {
...
+	keeperTestHelper.Require().NoError(err)
+
+	// allocate rewards to validator
+	keeperTestHelper.Ctx = keeperTestHelper.Ctx.WithBlockHeight(keeperTestHelper.Ctx.BlockHeight() + 1)
+	decTokens := sdk.DecCoins{{Denom: sdk.DefaultBondDenom, Amount: sdk.NewDec(20000)}}
+	keeperTestHelper.App.DistrKeeper.AllocateTokensToValidator(keeperTestHelper.Ctx, validator, decTokens)
+	keeperTestHelper.App.DistrKeeper.IncrementValidatorPeriod(keeperTestHelper.Ctx, validator)
+}
+
+func (keeperTestHelper *KeeperTestHelper) SetupGammPoolsWithBondDenomMultiplier(multipliers []sdk.Dec) []gammtypes.PoolI {
```
Comment (ValarDragon) [inline app/apptesting/test_suite.go:126]: Can we add a comment for this function? It wasn't clear to me from the name what exactly its aiming to create

File: app/apptesting/test_suite.go [hunk 7]
```diff
@@ -99,3 +106,100 @@ func (keeperTestHelper *KeeperTestHelper) EndBlock() {
...
+	keeperTestHelper.App.GAMMKeeper.SetParams(keeperTestHelper.Ctx, gammtypes.Params{
+		PoolCreationFee: sdk.Coins{},
+	})
+
+	bondDenom := keeperTestHelper.App.StakingKeeper.BondDenom(keeperTestHelper.Ctx)
+	acc1 := sdk.AccAddress(ed25519.GenPrivKey().PubKey().Address().Bytes())
```
Comment (ValarDragon) [inline app/apptesting/test_suite.go:132]: This isn't quite right (using tendermint not sdk crypto to generate an address)

Its fine for now, lets just add a TODO to use a standard helper method there?

**Review by ValarDragon — APPROVED** [review]
LGTM once comments are resolved!

**Comment by mattverse** [comment]
Changes made, planning to merge this as soon as lint and tests passes

**Pull request merged**